### ISSUE TITLE
tf.nn.softmax outputs negative values (equal to tf.nn.log_softmax)!

### ISSUE THREAD

**Issue by avati** [tensorflow/tensorflow#2315]
### Environment info
Operating System: Ubuntu 14.04.03

Installed version of CUDA and cuDNN: 
(please attach the output of `ls -l /path/to/cuda/lib/libcud*`):

ls -l /usr/local/cuda/lib*/libcud*
-rw-r--r-- 1 root root    322936 Apr 25 11:05 /usr/local/cuda/lib64/libcudadevrt.a
lrwxrwxrwx 1 root root        16 Apr 25 11:05 /usr/local/cuda/lib64/libcudart.so -> libcudart.so.7.5
lrwxrwxrwx 1 root root        19 Apr 25 11:05 /usr/local/cuda/lib64/libcudart.so.7.5 -> libcudart.so.7.5.18
-rwxr-xr-x 1 root root    383336 Apr 25 11:05 /usr/local/cuda/lib64/libcudart.so.7.5.18
-rw-r--r-- 1 root root    720192 Apr 25 11:05 /usr/local/cuda/lib64/libcudart_static.a
lrwxrwxrwx 1 3319 users       13 Feb  9 09:48 /usr/local/cuda/lib64/libcudnn.so -> libcudnn.so.4
lrwxrwxrwx 1 3319 users       17 Feb  9 09:48 /usr/local/cuda/lib64/libcudnn.so.4 -> libcudnn.so.4.0.7
-rwxrwxr-x 1 3319 users 61453024 Feb  8 14:12 /usr/local/cuda/lib64/libcudnn.so.4.0.7
-rw-rw-r-- 1 3319 users 62025862 Feb  8 14:12 /usr/local/cuda/lib64/libcudnn_static.a
-rw-r--r-- 1 root root    189170 Apr 25 11:05 /usr/local/cuda/lib/libcudadevrt.a
lrwxrwxrwx 1 root root        16 Apr 25 11:05 /usr/local/cuda/lib/libcudart.so -> libcudart.so.7.5
lrwxrwxrwx 1 root root        19 Apr 25 11:05 /usr/local/cuda/lib/libcudart.so.7.5 -> libcudart.so.7.5.18
-rwxr-xr-x 1 root root    311596 Apr 25 11:05 /usr/local/cuda/lib/libcudart.so.7.5.18
-rw-r--r-- 1 root root    558020 Apr 25 11:05 /usr/local/cuda/lib/libcudart_static.a

If installed from binary pip package, provide:

1. Which pip package you installed.

https://storage.googleapis.com/tensorflow/linux/gpu/tensorflow-0.7.1-cp27-none-linux_x86_64.whl

2. The output from python -c "import tensorflow; print(tensorflow.__version__)".

0.8.0rc0

If installed from sources, provide the commit hash:

### Steps to reproduce
It appears that tf.nn.softmax is output'ing negative values. After a lot of debugging, i added the following tf.Print line right after the call to tf.nn.softmax(). The output does indeed have negative values.. Strangely enough, the values match the output of tf.nn.log_softmax()!

https://github.com/sdlg/nlc/blob/04af660b23218c026983785f339afeea19cd4e25/nlc_model.py#L152

This issue could not be reproduced with simple test cases. tf.nn.softmax() behaves well in simple test cases.

### What have you tried?
1.

### Logs or other output that would be helpful
(If logs are large, please upload as attachment).

I tensorflow/core/kernels/logging_ops.cc:79] Checking for negative[-29.343697]
I tensorflow/core/kernels/logging_ops.cc:79] Checking for negative[-26.83346]
I tensorflow/core/kernels/logging_ops.cc:79] Checking for negative[-29.535496]
I tensorflow/core/kernels/logging_ops.cc:79] Checking for negative[-26.830235]
I tensorflow/core/kernels/logging_ops.cc:79] Checking for negative[-26.42857]

**Comment by girving** [comment]
Can you isolate an input to `tf.nn.softmax` that produces the problematic values?  It would be nice to have a simple test case that exhibits the problem.  Does adding a `tf.Print` to the inputs give you a reproducible bad input?

**Comment by avati** [comment]
I have been trying, but on simple graphs things seem to work fine. Whether I add tf.Print or not, my application always witnesses negative values (i.e behavior like tf.nn.log_softmax). It is very consistent in the erroneous behavior.

**Comment by girving** [comment]
If it's consistent, it seems like it's a deterministic problem in softmax.  If that's the case, you should be able to get a small reproducible instance via something like:

    ... lots of complicated graph construction code ...
    logits = ...
    probs = tf.nn.softmax(logits)
    ... more complicated graph construction code

    # Instead of
    np_output = sess.run(output)
    # do
    np_output, np_probs, np_logits = sess.run([output, probs, logits])

    # np_logits should now be a numpy array giving the reproducible bad input
    simple_prob = tf.nn.softmax(np_logits).eval()
    assert np.all(simple_prob == np_probs)

**Comment by girving** [comment]
Hmm, I found *a* bug, and it's certainly consistent with your excellent observation about `log_softmax`, but I'm not sure how you're triggering it.  Basically `tf.nn.softmax` and `tf.nn.log_softmax` check the op name to see if it starts with `"Log"`, but they should be checking the op type.

**Comment by girving** [comment]
`"Logistic"`!! :)

**Comment by girving** [comment]
Fix coming up.

**Comment by avati** [comment]
wow, no wonder i couldn't reproduce it in simple test cases, and it was consistent in the program!
